python/hail/ir/table_ir.py

```diff
@@ -0,0 +1,167 @@
+
+    def __str__(self):
+        return '(TableKeyBy ({}) {} {} {})'.format(
+            ' '.join([escape_id(x) for x in self.keys]), self.n_partitions, self.sort,
+            self.child)
+
+class TableMapRows(BaseIR):
+    def __init__(self, child, new_row, new_key, preserved_key_fields):
+        super().__init__()
+        self.child = child
+        self.new_row = new_row
+        self.new_key = new_key
+        self.preserved_key_fields = preserved_key_fields
+
+    def __str__(self):
+        return '(TableMapRows {} {} {} {})'.format(
+            ' '.join([escape_id(x) for x in self.new_key]) if self.new_key else 'None',
+            self.preserved_key_fields,
+            self.child, self.new_row)
+
+class TableUnkey(BaseIR):
+    def __init__(self, child):
+        super().__init__()
+        self.child = child
+
+    def __str__(self):
+        return '(TableUnkey {})'.format(self.child)
+
+class TableRead(BaseIR):
+    def __init__(self, path, drop_rows, typ):
+        super().__init__()
+        self.path = path
+        self.drop_rows = drop_rows
+        self.typ = typ
+
+    def __str__(self):
+        return '(TableRead "{}" {} {})'.format(
+            escape_str(self.path),
+            self.drop_rows,
+            self.typ)
+
+class TableImport(BaseIR):
+    def __init__(self, paths, typ, reader_options):
+        super().__init__()
+        self.paths = paths
+        self.typ = typ
+        self.reader_options = reader_options
+
+    def __str__(self):
+        return '(TableImport ({}) {} {})'.format(
+            ' '.join([escape_str(path) for path in self.paths]),
+            self.typ._jtype.parsableString(),
+            escape_str(json.dumps(self.reader_options)))
+
+class MatrixEntriesTable(BaseIR):
+    def __init__(self, child):
+        super().__init__()
+        self.child = child
+
+    def __str__(self):
+        return '(MatrixEntriesTable {})'.format(self.child)
+
+class TableFilter(BaseIR):
+    def __init__(self, child, pred):
+        super().__init__()
+        self.child = child
+        self.pred = pred
+
+    def __str__(self):
+        return '(TableFilter {} {})'.format(self.child, self.pred)
+
+class TableAggregateByKey(BaseIR):
+    def __init__(self, child, expr):
+        super().__init__()
+        self.child = child
+        self.expr = expr
+
+    def __str__(self):
+        return '(TableAggregateByKey {} {})'.format(self.child, self.expr)
+
+class MatrixColsTable(BaseIR):
+    def __init__(self, child):
+        super().__init__()
+        self.child = child
+
+    def __str__(self):
+        return '(MatrixColsTable {})'.format(self.child)
+
+class TableParallelize(BaseIR):
+    def __init__(self, typ, rows, n_partitions):
+        super().__init__()
+        self.typ = typ
+        self.rows = rows
+        self.n_partitions = n_partitions
+
+    def __str__(self):
+        return '(TableParallelize {} {} {})'.format(
+            self.typ,
+            self.rows,
```

Comment:
this could very well blow up the parser

Reply:
💣 💥

Reply:
How so?

Reply:
I'm a bit uncomfortable shoving the entirety of a tableparallelize inside this. We're sure the parser isn't doing backtracking, right? If not, then it's probably just a constant factor slower than the json parsers.

Reply:
Correct, I rewrote the string literal code (quotedLiteral, specifically) directly instead of going through the parser mechanism.  It makes a single pass and just handles quoting, it should be fast.  Then resulting string is passed to the usual JSON parser.